### PR TITLE
auth, etcdserver: permission of range requests

### PR DESCRIPTION
**Pull request by mitake**
Currently the auth mechanism doesn't support permissions of range
request. It just checks exact matching of key names even for range
queries. This commit adds a mechanism for setting permission to range
queries. Range queries are allowed if a range of the query is [begin1,
end1) and the user has a permission of reading [begin2, range2) and
[begin1, end2) is a subset of [begin2, range2). Range delete requests
will follow the same rule.

### CODE REVIEW

File: auth/authpb/auth.proto [hunk 1]
```diff
@@ -19,13 +19,14 @@ message User {
...
   bytes key = 1;
+  bytes range_end = 2;
 
   enum Type {
     READ = 0;
     WRITE = 1;
     READWRITE = 2;
   }
-  Type permType = 2;
+  Type permType = 3;
```
Comment (xiang90) [inline auth/authpb/auth.proto]: move the permType before key and range_end?

File: auth/store.go [hunk 2]
```diff
@@ -103,7 +103,7 @@ type AuthStore interface {
 	IsPutPermitted(header *pb.RequestHeader, key string) bool
 
 	// IsRangePermitted checks range permission of the user
-	IsRangePermitted(header *pb.RequestHeader, key string) bool
+	IsRangePermitted(header *pb.RequestHeader, key string, rangeEnd string) bool
```
Comment (xiang90) [inline auth/store.go]: key, rangeEnd string

File: auth/authpb/auth.proto (hunk 1 above)
Comment (mitake) [inline auth/authpb/auth.proto]: OK

File: auth/store.go (hunk 2 above)
Comment (mitake) [inline auth/store.go]: OK

File: auth/store.go [hunk 3]
```diff
@@ -569,7 +569,12 @@ func (as *authStore) RoleGrantPermission(r *pb.AuthRoleGrantPermissionRequest) (
 	return &pb.AuthRoleGrantPermissionResponse{}, nil
 }
 
-func (as *authStore) isOpPermitted(userName string, key string, write bool, read bool) bool {
+func isSubset(begin1, end1, begin2, end2 []byte) bool {
+	// return true if [begin1, end1) is a subset of [begin2, end2)
+	return 0 <= bytes.Compare(begin1, begin2) && bytes.Compare(begin2, end2) <= 0
```
Comment (gyuho) [inline auth/store.go]: Did you mean `bytes.Compare(end1, end2) <= 0`?

Otherwise, `end1 > end2` is fine?

File: auth/store.go [hunk 4]
```diff
@@ -607,18 +612,26 @@ func (as *authStore) isOpPermitted(userName string, key string, write bool, read
...
-					return true
+			if rangeEnd == "" {
+				if !bytes.Equal(perm.Key, []byte(key)) {
+					continue
 				}
-
-				if write && !read && perm.PermType == authpb.WRITE {
-					return true
+			} else {
+				if !isSubset([]byte(key), []byte(rangeEnd), perm.Key, perm.RangeEnd) {
```
Comment (xiang90) [inline auth/store.go]: if we have a perm from [a,b) and [b,d), [a,c) should be allowed. but now, it will be denied.

Comment (xiang90) [inline auth/store.go]: probably we need to keep a track of the accepted copy of the range. and keep on checking the unaccepted ones. if there is no unaccepted range left, then it is allowed.

File: auth/store.go (hunk 3 above)
Comment (mitake) [inline auth/store.go]: Ah, thanks for catch! I meant `bytes.Compare(end1, end2) <= 0`

File: auth/store.go (hunk 4 above)
Comment (mitake) [inline auth/store.go]: Yes... maybe unifying ranges at role granting would be simpler? How do you think?

**Comment by mitake** [comment]
@xiang90 updated for the unified range checking, could you take a look at the design? For testing the mechanism seriously, etcdctl should be changed for supporting range permission. I'll add it soon.

File: auth/range_perm_cache.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+	"github.com/coreos/etcd/auth/authpb"
+	"github.com/coreos/etcd/mvcc/backend"
+)
+
+type rangePerm struct {
+	begin, end string
+}
+
+type RangePermSlice []*rangePerm
```
Comment (xiang90) [inline auth/range_perm_cache.go]: RangePermSliceByBegin

Comment (xiang90) [inline auth/range_perm_cache.go]: also move this to the bottom of the file (we usually put utility funcs, structs at the end of the file)

File: auth/store.go [hunk 6]
```diff
@@ -592,22 +608,26 @@ func (as *authStore) isOpPermitted(userName string, key string, write bool, read
 		return false
 	}
 
-	for _, roleName := range user.Roles {
-		_, vs := tx.UnsafeRange(authRolesBucketName, []byte(roleName), nil, 0)
-		if len(vs) != 1 {
-			plog.Errorf("invalid role name %s for permission checking", roleName)
-			return false
-		}
+	if strings.Compare(rangeEnd, "") == 0 {
```
Comment (xiang90) [inline auth/store.go]: probably we can use similar mechanism for checking single key permission as range permission?

File: auth/range_perm_cache.go (hunk 5 above)
Comment (mitake) [inline auth/range_perm_cache.go]: OK. Should I move the definition of `unifiedRangePermissions` to the bottom, too?

File: auth/range_perm_cache.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+		}
+	}
+
+	return false
+}
+
+func (as *authStore) isRangeOpPermitted(tx backend.BatchTx, userName string, key, rangeEnd string, write, read bool) bool {
+	// assumption: tx is Lock()ed
+	_, ok := as.rangePermCache[userName]
+	if !ok {
```
Comment (xiang90) [inline auth/range_perm_cache.go]: let's do the ok case first. it is shorter. so we have less stuff in the indented code block.

File: auth/range_perm_cache.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+type unifiedRangePermissions struct {
+	// readPerms[i] and readPerms[j] (i != j) don't overlap
+	readPerms []*rangePerm
+	// writePerms[i] and writePerms[j] (i != j) don't overlap, too
+	writePerms []*rangePerm
+}
+
+func unifyPerms(perms []*rangePerm) []*rangePerm {
```
Comment (xiang90) [inline auth/range_perm_cache.go]: can we add a unit test for this algorithm?

File: auth/range_perm_cache.go (hunk 7 above)
Comment (mitake) [inline auth/range_perm_cache.go]: OK

File: auth/range_perm_cache.go (hunk 8 above)
Comment (mitake) [inline auth/range_perm_cache.go]: OK, I'll add the unit test

File: auth/store.go (hunk 6 above)
Comment (mitake) [inline auth/store.go]: OK, I'll try

File: auth/range_perm_cache.go (hunk 5 above)
Comment (xiang90) [inline auth/range_perm_cache.go]: sure. thanks!

**Comment by mitake** [comment]
@xiang90 updated for the comments expect the caching mechanism of single key permissions. It would make this PR too bloated so can I work on it on another PR?

File: auth/range_perm_cache.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+		}
+
+		i++
+	}
+
+	return ret
+}
+
+func (as *authStore) makeUnifiedPerms(tx backend.BatchTx, userName string) *unifiedRangePermissions {
+	_, vs := tx.UnsafeRange(authUsersBucketName, []byte(userName), nil, 0)
```
Comment (xiang90) [inline auth/range_perm_cache.go]: use getUser?

**Comment by xiang90** [comment]
@mitake Sure. OK. LGTM after fixing the getUser thing.

File: auth/range_perm_cache.go (hunk 9 above)
Comment (mitake) [inline auth/range_perm_cache.go]: Ah yes, thanks for pointing!

**Comment by mitake** [comment]
@xiang90 updated for getUser(), PTAL.

**Comment by mitake** [comment]
updated for style problems

**Comment by mitake** [comment]
@xiang90 I found some mistakes please don't merge this PR for now

**Comment by mitake** [comment]
@xiang90 resolve the problems and implemented range permission support in etcdctl, could you take a look?

File: etcdctl/ctlv3/command/role_command.go [hunk 10]
```diff
@@ -121,29 +121,42 @@ func roleGetCommandFunc(cmd *cobra.Command, args []string) {
...
 		}
 	}
 }
 
 // roleGrantPermissionCommandFunc executes the "role grant-permission" command.
 func roleGrantPermissionCommandFunc(cmd *cobra.Command, args []string) {
-	if len(args) != 3 {
-		ExitWithError(ExitBadArgs, fmt.Errorf("role grant command requires role name, permission type, and key as its argument."))
+	if len(args) < 3 {
+		ExitWithError(ExitBadArgs, fmt.Errorf("role grant command requires role name, permission type, and key (+ range end) as its argument."))
```
Comment (xiang90) [inline etcdctl/ctlv3/command/role_command.go]: update at line67 too?

Comment (xiang90) [inline etcdctl/ctlv3/command/role_command.go]: and key [endkey]

File: clientv3/auth.go [hunk 11]
```diff
@@ -78,13 +78,13 @@ type Auth interface {
 	RoleAdd(ctx context.Context, name string) (*AuthRoleAddResponse, error)
 
 	// RoleGrantPermission grants a permission to a role.
-	RoleGrantPermission(ctx context.Context, name string, key string, permType PermissionType) (*AuthRoleGrantPermissionResponse, error)
+	RoleGrantPermission(ctx context.Context, name string, key, rangeEnd string, permType PermissionType) (*AuthRoleGrantPermissionResponse, error)
 
 	// RoleGet gets a detailed information of a role.
 	RoleGet(ctx context.Context, role string) (*AuthRoleGetResponse, error)
 
 	// RoleRevokePermission revokes a key from a user.
```
Comment (xiang90) [inline clientv3/auth.go]: roleRevokePermission revokes a permission from a role.

File: clientv3/auth.go [hunk 12]
```diff
@@ -78,13 +78,13 @@ type Auth interface {
...
-	RoleRevokePermission(ctx context.Context, role string, key string) (*AuthRoleRevokePermissionResponse, error)
+	RoleRevokePermission(ctx context.Context, role string, key, rangeEnd string) (*AuthRoleRevokePermissionResponse, error)
```
Comment (xiang90) [inline clientv3/auth.go]: revoke probably should have a perm type too. we can resolve this in another pr.

Comment (mitake) [inline clientv3/auth.go]: OK, I'll do it in another PR

**Comment by mitake** [comment]
@xiang90 updated for the comments, PTAL

**Comment by mitake** [comment]
@xiang90 it seems that the test failure is the fleaky one, how do you think?

File: clientv3/auth.go [hunk 13]
```diff
@@ -78,13 +78,13 @@ type Auth interface {
...
 	// RoleGrantPermission grants a permission to a role.
-	RoleGrantPermission(ctx context.Context, name string, key string, permType PermissionType) (*AuthRoleGrantPermissionResponse, error)
+	RoleGrantPermission(ctx context.Context, name string, key, rangeEnd string, permType PermissionType) (*AuthRoleGrantPermissionResponse, error)
 
 	// RoleGet gets a detailed information of a role.
 	RoleGet(ctx context.Context, role string) (*AuthRoleGetResponse, error)
 
-	// RoleRevokePermission revokes a key from a user.
-	RoleRevokePermission(ctx context.Context, role string, key string) (*AuthRoleRevokePermissionResponse, error)
+	// RoleRevokePermission revokes a permission from a user.
```
Comment (xiang90) [inline clientv3/auth.go]: from a role not user.

**Comment by xiang90** [comment]
lgtm

File: clientv3/auth.go (hunk 13 above)
Comment (mitake) [inline clientv3/auth.go]: ahh, sorry for that

**Comment by mitake** [comment]
@xiang90 fixed the mistake in the comment, PTAL

**Pull request merged**